File: apps/ios/SignalASI/VoiceInteractionEvent.swift
import Foundation

enum VoiceInteractionEvent: Equatable {
  case capturePrepared(sessionId: String)
  case audioLevel(sessionId: String, rms: Float)
  case speechStarted(sessionId: String, atElapsedNs: Int64)
  case speechEnded(sessionId: String, atElapsedNs: Int64)
  case finalizationStarted(sessionId: String)
  case transcriptPartial(sessionId: String, value: TranscriptHypothesis)
  case transcriptStable(sessionId: String, value: TranscriptHypothesis)
  case transcriptFinal(sessionId: String, value: TranscriptHypothesis)
  case transcriptCorrected(sessionId: String, original: TranscriptHypothesis, corrected: TranscriptHypothesis)
  case routeSelected(sessionId: String, decision: VoiceRouteDecision)
  case localActionCompleted(sessionId: String)
  case modelDelta(sessionId: String, text: String)
  case agentAccepted(sessionId: String, runId: String)
  case agentProgress(sessionId: String, runId: String)
  case playbackStarted(sessionId: String, utteranceId: String)
  case completed(sessionId: String)
  case cancelled(sessionId: String, reasonCode: String)
  case failed(sessionId: String, failure: VoiceFailure)

  var sessionId: String {
    switch self {
    case let .capturePrepared(sessionId: sessionId),
         let .audioLevel(sessionId: sessionId, rms: _),
         let .speechStarted(sessionId: sessionId, atElapsedNs: _),
         let .speechEnded(sessionId: sessionId, atElapsedNs: _),
         let .finalizationStarted(sessionId: sessionId),
         let .transcriptPartial(sessionId: sessionId, value: _),
         let .transcriptStable(sessionId: sessionId, value: _),
         let .transcriptFinal(sessionId: sessionId, value: _),
         let .transcriptCorrected(sessionId: sessionId, original: _, corrected: _),
         let .routeSelected(sessionId: sessionId, decision: _),
         let .localActionCompleted(sessionId: sessionId),
         let .modelDelta(sessionId: sessionId, text: _),
         let .agentAccepted(sessionId: sessionId, runId: _),
         let .agentProgress(sessionId: sessionId, runId: _),
         let .playbackStarted(sessionId: sessionId, utteranceId: _),
         let .completed(sessionId: sessionId),
         let .cancelled(sessionId: sessionId, reasonCode: _),
         let .failed(sessionId: sessionId, failure: _):
      return sessionId
    }
  }

  var isTranscriptCorrection: Bool {
    if case .transcriptCorrected(sessionId: _, original: _, corrected: _) = self {
      return true
    }
    return false
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/voice/VoiceInteractionEvent.kt
package com.galaxyssi.chat.voice

sealed interface VoiceInteractionEvent {
    val sessionId: String

    data class CapturePrepared(override val sessionId: String) : VoiceInteractionEvent
    data class AudioLevel(override val sessionId: String, val rms: Float) : VoiceInteractionEvent
    data class SpeechStarted(override val sessionId: String, val atElapsedNs: Long) : VoiceInteractionEvent
    data class SpeechEnded(override val sessionId: String, val atElapsedNs: Long) : VoiceInteractionEvent
    data class FinalizationStarted(override val sessionId: String) : VoiceInteractionEvent
    data class TranscriptPartial(
        override val sessionId: String,
        val value: TranscriptHypothesis
    ) : VoiceInteractionEvent
    data class TranscriptStable(
        override val sessionId: String,
        val value: TranscriptHypothesis
    ) : VoiceInteractionEvent
    data class TranscriptFinal(
        override val sessionId: String,
        val value: TranscriptHypothesis
    ) : VoiceInteractionEvent
    data class TranscriptCorrected(
        override val sessionId: String,
        val original: TranscriptHypothesis,
        val corrected: TranscriptHypothesis
    ) : VoiceInteractionEvent
    data class RouteSelected(
        override val sessionId: String,
        val decision: VoiceRouteDecision
    ) : VoiceInteractionEvent
    data class LocalActionCompleted(override val sessionId: String) : VoiceInteractionEvent
    data class ModelDelta(override val sessionId: String, val text: String) : VoiceInteractionEvent
    data class AgentRunCreated(override val sessionId: String, val runId: String) : VoiceInteractionEvent
    data class AgentAccepted(override val sessionId: String, val runId: String) : VoiceInteractionEvent
    data class AgentProgress(override val sessionId: String, val runId: String) : VoiceInteractionEvent
    data class PlaybackStarted(override val sessionId: String, val utteranceId: String) : VoiceInteractionEvent
    data class Completed(override val sessionId: String) : VoiceInteractionEvent
    data class Cancelled(override val sessionId: String, val reasonCode: String) : VoiceInteractionEvent
    data class Failed(override val sessionId: String, val failure: VoiceFailure) : VoiceInteractionEvent
}
